### DISCUSSION TITLE
Is middleware supposed to be invoked twice for each request?

### DISCUSSION THREAD

**Discussion by madjam002** [cloudflare/next-on-pages#329] [Q&A] [answered]
I'm trying to write a middleware function that calls a remote endpoint if a certain route is hit. When deployed to Cloudflare though I can see that the endpoint is called twice from the middleware, I was then able to see the fact that middleware is invoked twice for each request even with a simple middleware like so:

```
export async function middleware(request: NextRequest) {
  console.log('got url', request.url)
}
```

I can see two "got url" log entries for each request in the Cloudflare Pages Functions debugger.

Is this expected behaviour?

EDIT: I've noticed that in .vercel/output/config.json there are two `"middlewarePath": "middleware",` entries:

```
    {
      "src": "^/.*$",
      "middlewareRawSrc": [],
      "middlewarePath": "middleware",
      "continue": true,
      "override": true
    },
...
    {
      "continue": true,
      "src": "^/.*$",
      "missing": [
        {
          "type": "header",
          "key": "x-prerender-revalidate",
          "value": "552e7405f9b64aa192d75556d0eea35b"
        }
      ],
      "middlewarePath": "middleware",
      "middlewareRawSrc": [
        "/:path*"
      ],
      "override": true
    },
```

If I remove the second one from the built bundle `(_worker.js/index.js)` then the middleware only gets called once and everything behaves as expected. Is there a reason there are two calls to the middleware here?

**Comment by james-elicx** [comment]
Hey there, thank you for raising this. It's an interesting point.

The config at `.vercel/output/config.json` is generated by Vercel - we then use it and run through it in our routing system. I wonder if perhaps we should add a check to see if the middleware has already been called during the running of that routing phase and, if so, avoid invoking it if it appears a second time.

If it's not too much trouble, do you know if your middleware is invoked twice when deployed to Vercel?

**Reply by madjam002** [reply]
I've just pushed the same build that @cloudflare/next-on-pages built using `yarn vercel deploy --prebuilt` and I can confirm that Vercel doesn't exhibit the same behaviour, the logs only appear once and only one request is made.

I double checked in the config.json file and there are still two entries to the middleware function.

**Reply by james-elicx** [reply, marked as answer]
Thank you for checking @madjam002, I appreciate it. I've opened a PR that I believe should resolve this problem, can you please try out the prerelease in https://github.com/cloudflare/next-on-pages/pull/337#issuecomment-1607257565?

**Reply by madjam002** [reply]
Thanks James, it's working perfectly now!

**Comment by dario-piotrowicz** [comment]
I'm closing the discussion as I believe it to be solved :slightly_smiling_face:
